{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  tady\n",
    "path_to_fna = 'E:\\VYSOKA\\BAKA\\data\\Botrytis_deweyae\\GCF_014898535.1_ASM1489853v1_genomic.fna'\n",
    "path_to_gff = 'E:\\VYSOKA\\BAKA\\data\\Botrytis_deweyae\\GCF_014898535.1_ASM1489853v1_genomic.gff'\n",
    "\n",
    "with open(path_to_gff) as f:\n",
    "    lines_gff = f.readlines()\n",
    "\n",
    "with open(path_to_fna) as f:\n",
    "    lines_fna = f.readlines()\n",
    "\n",
    "seq_regions = []\n",
    "nmbrofseq = 0\n",
    "CDS_in_region=[]\n",
    "for line in lines_gff:\n",
    "    if \"sequence-region\" in line:\n",
    "        nmbrofseq = nmbrofseq +1\n",
    "        seq_regions.append(CDS_in_region)\n",
    "        CDS_in_region = []\n",
    "    if \"CDS\" in line:\n",
    "        l = line.split('\\t')[3:9]\n",
    "        l[5] = l[5].split(';')[0]\n",
    "        l[0] = int(l[0])\n",
    "        l[1] = int(l[1])\n",
    "        CDS_in_region.append(l)\n",
    "\n",
    "\n",
    "    if \"###\" in line.split(\" \")[0]:\n",
    "        seq_regions.append(CDS_in_region)\n",
    "\n",
    "seq_regions.pop(0)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def sortfucn(e):\n",
    "    return e[0]\n",
    "\n",
    "for i in range(0,len(seq_regions)-1):\n",
    "    if len(seq_regions[i]) ==0 :\n",
    "         continue\n",
    "    seq_regions[i].sort(key=sortfucn)\n",
    "\n",
    "AllCDS = seq_regions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\item  Aaosphaeria_arxii\n",
      "\\item  Acaromyces_ingoldii\n",
      "\\item  Agaricus_bisporus\n",
      "\\item  Alternaria_alternata\n",
      "\\item  Alternaria_arborescens\n",
      "\\item  Alternaria_arbusti\n",
      "\\item  Alternaria_atra\n",
      "\\item  Alternaria_burnsii\n",
      "\\item  Alternaria_conjuncta\n",
      "\\item  Alternaria_ethzedia\n",
      "\\item  Alternaria_hordeiaustralica\n",
      "\\item  Alternaria_incomplexa\n",
      "\\item  Alternaria_infectoria\n",
      "\\item  Alternaria_metachromatica\n",
      "\\item  Alternaria_novae-zelandiae\n",
      "\\item  Alternaria_postmessia\n",
      "\\item  Alternaria_rosae\n",
      "\\item  Alternaria_triticimaculans\n",
      "\\item  Alternaria_ventricosa\n",
      "\\item  Alternaria_viburni\n",
      "\\item  Amorphotheca_resinae\n",
      "\\item  Annulohypoxylon_maeteangense\n",
      "\\item  Annulohypoxylon_truncatum\n",
      "\\item  Apiotrichum_porosum\n",
      "\\item  Aplosporella_prunicola\n",
      "\\item  Arthroderma_uncinatum\n",
      "\\item  Ascochyta_rabiei\n",
      "\\item  Ascoidea_rubescens\n",
      "\\item  Aspergillus_aculeatinus\n",
      "\\item  Aspergillus_aculeatus\n",
      "\\item  Aspergillus_alliaceus\n",
      "\\item  Aspergillus_bombycis\n",
      "\\item  Aspergillus_brunneoviolaceus\n",
      "\\item  Aspergillus_caelatus\n",
      "\\item  Aspergillus_campestris\n",
      "\\item  Aspergillus_candidus\n",
      "\\item  Aspergillus_chevalieri\n",
      "\\item  Aspergillus_clavatus\n",
      "\\item  Aspergillus_costaricensis\n",
      "\\item  Aspergillus_eucalypticola\n",
      "\\item  Aspergillus_fijiensis\n",
      "\\item  Aspergillus_fischeri\n",
      "\\item  Aspergillus_flavus\n",
      "\\item  Aspergillus_fumigatus\n",
      "\\item  Aspergillus_glaucus\n",
      "\\item  Aspergillus_heteromorphus\n",
      "\\item  Aspergillus_homomorphus\n",
      "\\item  Aspergillus_ibericus\n",
      "\\item  Aspergillus_japonicus\n",
      "\\item  Aspergillus_lentulus\n",
      "\\item  Aspergillus_luchuensis\n",
      "\\item  Aspergillus_melleus\n",
      "\\item  Aspergillus_mulundensis\n",
      "\\item  Aspergillus_neoniger\n",
      "\\item  Aspergillus_nidulans\n",
      "\\item  Aspergillus_niger\n",
      "\\item  Aspergillus_nomiae\n",
      "\\item  Aspergillus_novofumigatus\n",
      "\\item  Aspergillus_ochraceoroseus\n",
      "\\item  Aspergillus_oryzae\n",
      "\\item  Aspergillus_piperis\n",
      "\\item  Aspergillus_pseudonomiae\n",
      "\\item  Aspergillus_pseudotamarii\n",
      "\\item  Aspergillus_pseudoviridinutans\n",
      "\\item  Aspergillus_puulaauensis\n",
      "\\item  Aspergillus_ruber\n",
      "\\item  Aspergillus_saccharolyticus\n",
      "\\item  Aspergillus_sclerotioniger\n",
      "\\item  Aspergillus_steynii\n",
      "\\item  Aspergillus_sydowii\n",
      "\\item  Aspergillus_tanneri\n",
      "\\item  Aspergillus_terreus\n",
      "\\item  Aspergillus_thermomutatus\n",
      "\\item  Aspergillus_tubingensis\n",
      "\\item  Aspergillus_udagawae\n",
      "\\item  Aspergillus_uvarum\n",
      "\\item  Aspergillus_vadensis\n",
      "\\item  Aspergillus_versicolor\n",
      "\\item  Aspergillus_viridinutans\n",
      "\\item  Aspergillus_welwitschiae\n",
      "\\item  Aspergillus_wentii\n",
      "\\item  Aureobasidium_melanogenum\n",
      "\\item  Aureobasidium_namibiae\n",
      "\\item  Aureobasidium_pullulans\n",
      "\\item  Aureobasidium_subglaciale\n",
      "\\item  Auricularia_subglabra\n",
      "\\item  Babjeviella_inositovora\n",
      "\\item  Bacidia_gigantensis\n",
      "\\item  Batrachochytrium_dendrobatidis\n",
      "\\item  Baudoinia_panamericana\n",
      "\\item  Beauveria_bassiana\n",
      "\\item  Bipolaris_maydis\n",
      "\\item  Bipolaris_oryzae\n",
      "\\item  Bipolaris_sorokiniana\n",
      "\\item  Bipolaris_victoriae\n",
      "\\item  Bipolaris_zeicola\n",
      "\\item  Blastomyces_dermatitidis\n",
      "\\item  Blastomyces_gilchristii\n",
      "\\item  Boeremia_exigua\n",
      "\\item  Botrytis_byssoidea\n",
      "\\item  Botrytis_cinerea\n",
      "\\item  Botrytis_deweyae\n",
      "\\item  Botrytis_fragariae\n",
      "\\item  Botrytis_porri\n",
      "\\item  Botrytis_sinoallii\n",
      "\\item  Brettanomyces_bruxellensis\n",
      "\\item  Brettanomyces_nanus\n",
      "\\item  Candida_albicans\n",
      "\\item  Candida_dubliniensis\n",
      "\\item  Candida_jiufengensis\n",
      "\\item  Candida_margitis\n",
      "\\item  Candida_orthopsilosis\n",
      "\\item  Candida_oxycetoniae\n",
      "\\item  Candida_parapsilosis\n",
      "\\item  Candida_pseudojiufengensis\n",
      "\\item  Candida_theae\n",
      "\\item  Candida_tropicalis\n",
      "\\item  Candida_viswanathii\n",
      "\\item  Cantharellus_anzutake\n",
      "\\item  Capronia_coronata\n",
      "\\item  Capronia_epimyces\n",
      "\\item  Ceraceosorus_guamensis\n",
      "\\item  Cercospora_beticola\n",
      "\\item  Cercospora_kikuchii\n",
      "\\item  Chaetomium_globosum\n",
      "\\item  Cladophialophora_bantiana\n",
      "\\item  Cladophialophora_carrionii\n",
      "\\item  Cladophialophora_immunda\n",
      "\\item  Cladophialophora_psammophila\n",
      "\\item  Cladophialophora_yegresii\n",
      "\\item  Clavispora_lusitaniae\n",
      "\\item  Coccidioides_immitis\n",
      "\\item  Coccidioides_posadasii\n",
      "\\item  Cokeromyces_recurvatus\n",
      "\\item  Colletotrichum_aenigma\n",
      "\\item  Colletotrichum_fioriniae\n",
      "\\item  Colletotrichum_fructicola\n",
      "\\item  Colletotrichum_gloeosporioides\n",
      "\\item  Colletotrichum_graminicola\n",
      "\\item  Colletotrichum_higginsianum\n",
      "\\item  Colletotrichum_karsti\n",
      "\\item  Colletotrichum_lupini\n",
      "\\item  Colletotrichum_orchidophilum\n",
      "\\item  Colletotrichum_scovillei\n",
      "\\item  Colletotrichum_siamense\n",
      "\\item  Colletotrichum_spaethianum\n",
      "\\item  Colletotrichum_truncatum\n",
      "\\item  Coniophora_puteana\n",
      "\\item  Coniosporium_apollinis\n",
      "\\item  Coprinopsis_cinerea\n",
      "\\item  Cordyceps_fumosorosea\n",
      "\\item  Cordyceps_militaris\n",
      "\\item  Cryphonectria_parasitica\n",
      "\\item  Cryptococcus_amylolentus\n",
      "\\item  Cryptococcus_gattii_VGI\n",
      "\\item  Cryptococcus_neoformans\n",
      "\\item  Cryptococcus_wingfieldii\n",
      "\\item  Cucurbitaria_berberidis\n",
      "\\item  Cutaneotrichosporon_oleaginosum\n",
      "\\item  Cyberlindnera_jadinii\n",
      "\\item  Cyphellophora_europaea\n",
      "\\item  Dacryopinax_primogenitus\n",
      "\\item  Dactylellina_haptotyla\n",
      "\\item  Daldinia_caldariorum\n",
      "\\item  Daldinia_childiae\n",
      "\\item  Daldinia_decipiens\n",
      "\\item  Daldinia_loculata\n",
      "\\item  Daldinia_vernicosa\n",
      "\\item  Debaryomyces_fabryi\n",
      "\\item  Debaryomyces_hansenii\n",
      "\\item  Diaporthe_batatas\n",
      "\\item  Diaporthe_citri\n",
      "\\item  Dichomitus_squalens\n",
      "\\item  Didymella_exigua\n",
      "\\item  Diplodia_corticola\n",
      "\\item  Dissoconium_aciculare\n",
      "\\item  Diutina_rugosa\n",
      "\\item  Dothidotthia_symphoricarpi\n",
      "\\item  Drechmeria_coniospora\n",
      "\\item  Drepanopeziza_brunnea\n",
      "\\item  Durotheca_rogersii\n",
      "\\item  Emericellopsis_atlantica\n",
      "\\item  Emericellopsis_cladophorae\n",
      "\\item  Encephalitozoon_cuniculi\n",
      "\\item  Encephalitozoon_hellem\n",
      "\\item  Encephalitozoon_intestinalis\n",
      "\\item  Encephalitozoon_romaleae\n",
      "\\item  Endocarpon_pusillum\n",
      "\\item  Enterocytozoon_bieneusi\n",
      "\\item  Epithele_typhae\n",
      "\\item  Eremomyces_bilateralis\n",
      "\\item  Eremothecium_cymbalariae\n",
      "\\item  Eremothecium_gossypii\n",
      "\\item  Eremothecium_sinecaudum\n",
      "\\item  Eutypa_lata\n",
      "\\item  Exophiala_aquamarina\n",
      "\\item  Exophiala_dermatitidis\n",
      "\\item  Exophiala_mesophila\n",
      "\\item  Exophiala_oligosperma\n",
      "\\item  Exophiala_spinifera\n",
      "\\item  Exophiala_xenobiotica\n",
      "\\item  Exserohilum_turcicum\n",
      "\\item  Fibroporia_radiculosa\n",
      "\\item  Filobasidium_floriforme\n",
      "\\item  Fomitiporia_mediterranea\n",
      "\\item  Fonsecaea_erecta\n",
      "\\item  Fonsecaea_monophora\n",
      "\\item  Fonsecaea_multimorphosa\n",
      "\\item  Fonsecaea_nubica\n",
      "\\item  Fonsecaea_pedrosoi\n",
      "\\item  Fulvia_fulva\n",
      "\\item  Fusarium_coffeatum\n",
      "\\item  Fusarium_flagelliforme\n",
      "\\item  Fusarium_fujikuroi\n",
      "\\item  Fusarium_graminearum\n",
      "\\item  Fusarium_mangiferae\n",
      "\\item  Fusarium_musae\n",
      "\\item  Fusarium_odoratissimum\n",
      "\\item  Fusarium_oxysporum\n",
      "\\item  Fusarium_poae\n",
      "\\item  Fusarium_proliferatum\n",
      "\\item  Fusarium_pseudograminearum\n",
      "\\item  Fusarium_redolens\n",
      "\\item  Fusarium_solani\n",
      "\\item  Fusarium_subglutinans\n",
      "\\item  Fusarium_tjaetaba\n",
      "\\item  Fusarium_vanettenii\n",
      "\\item  Fusarium_venenatum\n",
      "\\item  Fusarium_verticillioides\n",
      "\\item  Gaeumannomyces_tritici\n",
      "\\item  Gamsiella_multidivaricata\n",
      "\\item  Geosmithia_morbida\n",
      "\\item  Gilbertella_persicaria\n",
      "\\item  Glarea_lozoyensis\n",
      "\\item  Gloeophyllum_trabeum\n",
      "\\item  Grosmannia_clavigera\n",
      "\\item  Guyanagaster_necrorhizus\n",
      "\\item  Halteromyces_radiatus\n",
      "\\item  Heterobasidion_irregulare\n",
      "\\item  Hirsutella_rhossiliensis\n",
      "\\item  Histoplasma_capsulatum\n",
      "\\item  Histoplasma_mississippiense_nom._inval._\n",
      "\\item  Hyaloscypha_bicolor\n",
      "\\item  Hyphopichia_burtonii\n",
      "\\item  Hypoxylon_fragiforme\n",
      "\\item  Hypoxylon_trugodes\n",
      "\\item  Ilyonectria_robusta\n",
      "\\item  Jaminaea_rosea\n",
      "\\item  Kalmanozyma_brasiliensis\n",
      "\\item  Kazachstania_africana\n",
      "\\item  Kazachstania_barnettii\n",
      "\\item  Kazachstania_naganishii\n",
      "\\item  Kickxella_alabastrina\n",
      "\\item  Kluyveromyces_lactis\n",
      "\\item  Kluyveromyces_marxianus\n",
      "\\item  Kockovaella_imperatae\n",
      "\\item  Komagataella_phaffii\n",
      "\\item  Kuraishia_capsulata\n",
      "\\item  Kwoniella_bestiolae\n",
      "\\item  Kwoniella_dejecticola\n",
      "\\item  Kwoniella_mangrovensis\n",
      "\\item  Kwoniella_pini\n",
      "\\item  Kwoniella_shandongensis\n",
      "\\item  Laccaria_bicolor\n",
      "\\item  Lachancea_lanzarotensis\n",
      "\\item  Lachancea_thermotolerans\n",
      "\\item  Lachnellula_hyalina\n",
      "\\item  Laetiporus_sulphureus\n",
      "\\item  Lasiodiplodia_theobromae\n",
      "\\item  Lentinula_edodes\n",
      "\\item  Leptosphaeria_maculans\n",
      "\\item  Letharia_columbiana\n",
      "\\item  Letharia_lupina\n",
      "\\item  Linderina_pennispora\n",
      "\\item  Lindgomyces_ingoldianus\n",
      "\\item  Lobosporangium_transversale\n",
      "\\item  Lodderomyces_elongisporus\n",
      "\\item  Macroventuria_anomochaeta\n",
      "\\item  Malassezia_globosa\n",
      "\\item  Malassezia_pachydermatis\n",
      "\\item  Malassezia_restricta\n",
      "\\item  Malassezia_sympodialis\n",
      "\\item  Marasmius_oreades\n",
      "\\item  Meira_miltonrushii\n",
      "\\item  Melampsora_laricis-populina\n",
      "\\item  Metarhizium_acridum\n",
      "\\item  Metarhizium_album\n",
      "\\item  Metarhizium_brunneum\n",
      "\\item  Metarhizium_majus\n",
      "\\item  Metarhizium_robertsii\n",
      "\\item  Metschnikowia_bicuspidata\n",
      "\\item  Meyerozyma_guilliermondii\n",
      "\\item  Microdochium_trichocladiopsis\n",
      "\\item  Microsporum_canis\n",
      "\\item  Millerozyma_farinosa\n",
      "\\item  Mitosporidium_daphniae\n",
      "\\item  Mixia_osmundae\n",
      "\\item  Moesziomyces_antarcticus\n",
      "\\item  Mollisia_scopiformis\n",
      "\\item  Moniliophthora_perniciosa\n",
      "\\item  Moniliophthora_roreri\n",
      "\\item  Morchella_importuna\n",
      "\\item  Morchella_sextelata\n",
      "\\item  Mucor_mucedo\n",
      "\\item  Mycena_indigotica\n",
      "\\item  Mytilinidion_resinicola\n",
      "\\item  Nannizzia_gypsea\n",
      "\\item  Naumovozyma_castellii\n",
      "\\item  Naumovozyma_dairenensis\n",
      "\\item  Nematocida_homosporus\n",
      "\\item  Nematocida_major\n",
      "\\item  Nematocida_minor\n",
      "\\item  Nematocida_parisii\n",
      "\\item  Neoantrodia_serialis\n",
      "\\item  Neoarthrinium_moseri\n",
      "\\item  Neofusicoccum_parvum\n",
      "\\item  Neohortaea_acidophila\n",
      "\\item  Neurospora_crassa\n",
      "\\item  Neurospora_tetrasperma\n",
      "\\item  Nosema_ceranae\n",
      "\\item  Ogataea_angusta\n",
      "\\item  Ogataea_haglerorum\n",
      "\\item  Ogataea_parapolymorpha\n",
      "\\item  Ogataea_philodendri\n",
      "\\item  Ogataea_polymorpha\n",
      "\\item  Ophidiomyces_ophidiicola\n",
      "\\item  Orbilia_oligospora\n",
      "\\item  Ordospora_colligata\n",
      "\\item  Paecilomyces_variotii\n",
      "\\item  Paracoccidioides_brasiliensis\n",
      "\\item  Paracoccidioides_lutzii\n",
      "\\item  Paraphaeosphaeria_sporulosa\n",
      "\\item  Parastagonospora_nodorum\n",
      "\\item  Penicilliopsis_zonata\n",
      "\\item  Penicillium_arizonense\n",
      "\\item  Penicillium_digitatum\n",
      "\\item  Penicillium_expansum\n",
      "\\item  Penicillium_griseofulvum\n",
      "\\item  Penicillium_oxalicum\n",
      "\\item  Penicillium_roqueforti\n",
      "\\item  Penicillium_rubens\n",
      "\\item  Penicillium_solitum\n",
      "\\item  Pestalotiopsis_fici\n",
      "\\item  Phaeoacremonium_minimum\n",
      "\\item  Phanerochaete_carnosa\n",
      "\\item  Phialophora_attinorum\n",
      "\\item  Phycomyces_blakesleeanus\n",
      "\\item  Pichia_kudriavzevii\n",
      "\\item  Pichia_membranifaciens\n",
      "\\item  Pisolithus_orientalis\n",
      "\\item  Pleurotus_ostreatus\n",
      "\\item  Pneumocystis_carinii\n",
      "\\item  Pneumocystis_jirovecii\n",
      "\\item  Pneumocystis_murina\n",
      "\\item  Pochonia_chlamydosporia\n",
      "\\item  Podospora_anserina\n",
      "\\item  Postia_placenta\n",
      "\\item  Protomyces_lactucae-debilis\n",
      "\\item  Pseudocercospora_fijiensis\n",
      "\\item  Pseudogymnoascus_destructans\n",
      "\\item  Pseudogymnoascus_verrucosus\n",
      "\\item  Pseudomassariella_vexata\n",
      "\\item  Pseudomicrostroma_glucosiphilum\n",
      "\\item  Pseudovirgaria_hyperparasitica\n",
      "\\item  Pseudozyma_flocculosa\n",
      "\\item  Pseudozyma_hubeiensis\n",
      "\\item  Psilocybe_cubensis\n",
      "\\item  Puccinia_graminis\n",
      "\\item  Puccinia_striiformis\n",
      "\\item  Punctularia_strigosozonata\n",
      "\\item  Purpureocillium_lilacinum\n",
      "\\item  Purpureocillium_takamizusanense\n",
      "\\item  Pyrenophora_teres\n",
      "\\item  Pyrenophora_tritici-repentis\n",
      "\\item  Pyricularia_grisea\n",
      "\\item  Pyricularia_oryzae\n",
      "\\item  Pyricularia_pennisetigena\n",
      "\\item  Radiomyces_spectabilis\n",
      "\\item  Ramularia_collo-cygni\n",
      "\\item  Rasamsonia_emersonii\n",
      "\\item  Rhinocladiella_mackenziei\n",
      "\\item  Rhizoctonia_solani\n",
      "\\item  Rhizophagus_irregularis\n",
      "\\item  Rhizopus_microsporus\n",
      "\\item  Rhodofomes_roseus\n",
      "\\item  Rhodotorula_graminis\n",
      "\\item  Rhodotorula_toruloides\n",
      "\\item  Saccharomyces_arboricola\n",
      "\\item  Saccharomyces_cerevisiae\n",
      "\\item  Saccharomyces_eubayanus\n",
      "\\item  Saccharomyces_paradoxus\n",
      "\\item  Saccharomycodes_ludwigii\n",
      "\\item  Saitoella_complicata\n",
      "\\item  Saprochaete_ingens\n",
      "\\item  Scedosporium_apiospermum\n",
      "\\item  Scheffersomyces_spartinae\n",
      "\\item  Scheffersomyces_stipitis\n",
      "\\item  Schizophyllum_commune\n",
      "\\item  Schizosaccharomyces_cryophilus\n",
      "\\item  Schizosaccharomyces_japonicus\n",
      "\\item  Schizosaccharomyces_octosporus\n",
      "\\item  Schizosaccharomyces_pombe\n",
      "\\item  Sclerotinia_sclerotiorum\n",
      "\\item  Serpula_lacrymans\n",
      "\\item  Sodiomyces_alkalinus\n",
      "\\item  Sordaria_macrospora\n",
      "\\item  Sparassis_crispa\n",
      "\\item  Spathaspora_passalidarum\n",
      "\\item  Sphaerulina_musiva\n",
      "\\item  Spizellomyces_punctatus\n",
      "\\item  Sporisorium_graminicola\n",
      "\\item  Sporothrix_brasiliensis\n",
      "\\item  Sporothrix_schenckii\n",
      "\\item  Stereum_hirsutum\n",
      "\\item  Sugiyamaella_lignohabitans\n",
      "\\item  Suhomyces_tanzawaensis\n",
      "\\item  Suillus_bovinus\n",
      "\\item  Suillus_clintonianus\n",
      "\\item  Suillus_discolor\n",
      "\\item  Suillus_fuscotomentosus\n",
      "\\item  Suillus_paluster\n",
      "\\item  Suillus_plorans\n",
      "\\item  Suillus_subalutaceus\n",
      "\\item  Suillus_subaureus\n",
      "\\item  Synchytrium_microbalum\n",
      "\\item  Talaromyces_amestolkiae\n",
      "\\item  Talaromyces_atroroseus\n",
      "\\item  Talaromyces_marneffei\n",
      "\\item  Talaromyces_proteolyticus\n",
      "\\item  Talaromyces_rugulosus\n",
      "\\item  Talaromyces_stipitatus\n",
      "\\item  Tetrapisispora_blattae\n",
      "\\item  Tetrapisispora_phaffii\n",
      "\\item  Thermochaetoides_thermophila\n",
      "\\item  Thermothelomyces_thermophilus\n",
      "\\item  Thermothielavioides_terrestris\n",
      "\\item  Thyridium_curvatum\n",
      "\\item  Tilletiaria_anomala\n",
      "\\item  Tilletiopsis_washingtonensis\n",
      "\\item  Torulaspora_delbrueckii\n",
      "\\item  Torulaspora_globosa\n",
      "\\item  Trametes_versicolor\n",
      "\\item  Trematosphaeria_pertusa\n",
      "\\item  Tremella_mesenterica\n",
      "\\item  Tricharina_praecox\n",
      "\\item  Trichoderma_asperellum\n",
      "\\item  Trichoderma_atroviride\n",
      "\\item  Trichoderma_citrinoviride\n",
      "\\item  Trichoderma_gamsii\n",
      "\\item  Trichoderma_harzianum\n",
      "\\item  Trichoderma_reesei\n",
      "\\item  Trichoderma_virens\n",
      "\\item  Trichophyton_benhamiae\n",
      "\\item  Trichophyton_rubrum\n",
      "\\item  Trichophyton_verrucosum\n",
      "\\item  Trichosporon_asahii\n",
      "\\item  Truncatella_angustata\n",
      "\\item  Tuber_melanosporum\n",
      "\\item  Umbelopsis_ramanniana\n",
      "\\item  Uncinocarpus_reesii\n",
      "\\item  Ustilaginoidea_virens\n",
      "\\item  Ustilago_hordei\n",
      "\\item  Ustilago_maydis\n",
      "\\item  Vanderwaltozyma_polyspora\n",
      "\\item  Vavraia_culicis\n",
      "\\item  Venustampulla_echinocandica\n",
      "\\item  Verruconis_gallopava\n",
      "\\item  Verticillium_alfalfae\n",
      "\\item  Verticillium_dahliae\n",
      "\\item  Verticillium_nonalfalfae\n",
      "\\item  Vittaforma_corneae\n",
      "\\item  Wallemia_ichthyophaga\n",
      "\\item  Wallemia_mellicola\n",
      "\\item  Westerdykella_ornata\n",
      "\\item  Wickerhamiella_sorbophila\n",
      "\\item  Wickerhamomyces_anomalus\n",
      "\\item  Wickerhamomyces_ciferrii\n",
      "\\item  Xylaria_bambusicola\n",
      "\\item  Xylona_heveae\n",
      "\\item  Yamadazyma_tenuis\n",
      "\\item  Yarrowia_lipolytica\n",
      "\\item  Zasmidium_cellare\n",
      "\\item  Zygosaccharomyces_rouxii\n",
      "\\item  Zygotorulaspora_mrakii\n",
      "\\item  Zymoseptoria_tritici\n",
      "\\item  _Candida_auris\n",
      "\\item  _Candida_duobushaemulonis\n",
      "\\item  _Candida_glabrata\n",
      "\\item  _Candida_haemuloni\n",
      "\\item  _Candida_pseudohaemulonii\n",
      "\\item  _Candida_subhashii\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "dir_list = os.listdir(\"E:/VYSOKA/BAKA/data\")\n",
    "for houba in dir_list:\n",
    "    print('\\item ', houba)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "seq_regions[25][191]\n",
    "for region in seq_regions:\n",
    "    print(\"---------------------------\")\n",
    "    for cds in region:\n",
    "        print(cds[0],cds[5])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy\n",
    "chDuplicates = copy.deepcopy(seq_regions)\n",
    "for region in chDuplicates :\n",
    "    for cds in region:\n",
    "        chDuplicates[chDuplicates.index(region)][region.index(cds)] = cds[0]\n",
    "\n",
    "\n",
    "IndexesTodelete = []\n",
    "for region in chDuplicates:\n",
    "    indexes = []\n",
    "    newregion = []\n",
    "    for cds in region:\n",
    "        if cds not in newregion: \n",
    "            newregion.append(cds)\n",
    "            continue\n",
    "        index = [i for i, x in enumerate(region) if x == cds]\n",
    "        for ind in index:\n",
    "            if ind in indexes:continue\n",
    "            indexes.append(ind)\n",
    "    IndexesTodelete.append(indexes)\n",
    "\n",
    "for region in IndexesTodelete:\n",
    "    deletedNumbers = 0\n",
    "    for index in region:\n",
    "        try:\n",
    "            del seq_regions[IndexesTodelete.index(region)][index- deletedNumbers]\n",
    "        except:\n",
    "            deletedNumbers = deletedNumbers + 1\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = []\n",
    "indexes = [i for i, x in enumerate([1, 2, 1]) if x == 1]\n",
    "print(indexes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promotors = []\n",
    "\n",
    "for region in seq_regions:\n",
    "    promotors_in_region = []\n",
    "    for CDS in region:\n",
    "        if CDS[3] == '+' :\n",
    "            promotors_in_region.append([int(CDS[0])-1000,int(CDS[0])-1,\"+\",'n',CDS[5]])\n",
    "            continue\n",
    "        promotors_in_region.append([int(CDS[1])+1,int(CDS[1])+1000,\"-\",'n',CDS[5]])\n",
    "    promotors.append(promotors_in_region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(seq_regions[23][145:149])\n",
    "print(seq_regions[22][21][0:5] in seq_regions[22][20:30][0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for region in promotors:\n",
    "    for promotor in region:\n",
    "        if promotor[2] == '+' :\n",
    "            if region.index(promotor) == 0 : continue\n",
    "            secondIndexOfCDSbefore = int(seq_regions[promotors.index(region)][region.index(promotor)-1][1])\n",
    "            if promotor[0] <= secondIndexOfCDSbefore :\n",
    "                promotor[3] = int(promotor[1]) - secondIndexOfCDSbefore\n",
    "                if promotor[3] < 0 :\n",
    "                    promotor[3] = 0\n",
    "            continue\n",
    "        if region.index(promotor)+1 == len(seq_regions[promotors.index(region)]) : continue\n",
    "        firtsIndexOfCDSafter = int(seq_regions[promotors.index(region)][region.index(promotor)+1][0])\n",
    "        if promotor[1] >= firtsIndexOfCDSafter :\n",
    "            promotor[3] = firtsIndexOfCDSafter - promotor[0]\n",
    "            if promotor[3] < 0 : promotor[3] = 0\n",
    "\n",
    "\n",
    "            \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import intervals as I\n",
    "\n",
    "Numbers = []\n",
    "for region in seq_regions:\n",
    "    nbr_in_region = []\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        nbr_in_region.append(I.closed(number1,number2))\n",
    "    Numbers.append(nbr_in_region)\n",
    "\n",
    "\n",
    "for region in promotors :\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        Numbers[promotors.index(region)].append(I.closed(number1,number2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# moc na dlouho\n",
    "region_ranges = []\n",
    "offgene = []\n",
    "\n",
    "for line in lines_gff:\n",
    "    if \"sequence-region\" in line:\n",
    "        number1 =int(line.split(' ')[2])\n",
    "        number2= int(line.split(' ')[3].replace(\"\\n\",''))\n",
    "        region_ranges.append([number1,number2])\n",
    "\n",
    "for region in region_ranges:\n",
    "    nmbrs = list(range(region[0],region[1]))\n",
    "    r = []\n",
    "    for n in nmbrs :\n",
    "        IsInItrevral = False\n",
    "        for interval in Numbers[region_ranges.index(region)] :\n",
    "            if n in interval : \n",
    "                IsInItrevral = True\n",
    "                break\n",
    "        if IsInItrevral : continue\n",
    "        r.append(n)\n",
    "    offgene.append(r)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#stara verze\n",
    "Numbers = []\n",
    "for region in seq_regions:\n",
    "    nbr_in_region = []\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        nbr_in_region.append(list(range(number1,number2)))\n",
    "    Numbers.append(nbr_in_region)\n",
    "\n",
    "for region in promotors :\n",
    "    nbr_in_region = []\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        Numbers[promotors.index(region)].append((list(range(number1,number2))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Allnumbers = []\n",
    "for region in Numbers :\n",
    "    wholestring = []\n",
    "    for nbr in region:\n",
    "        wholestring.append(str(nbr))\n",
    "    string =''.join(wholestring)\n",
    "    res = string.strip('][').replace('][',',').split(',')\n",
    "    Allnumbers.append(res)\n",
    "\n",
    "Allnumbers = [[ subelt for subelt in elt if subelt != '' ] for elt in Allnumbers]\n",
    "\n",
    "IndexesOfOffgene = []\n",
    "for region in Allnumbers:\n",
    "    res = [int(i) for i in region]\n",
    "    IndexesOfOffgene.append(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_ranges = []\n",
    "offgene = []\n",
    "\n",
    "for line in lines_gff:\n",
    "    if \"sequence-region\" in line:\n",
    "        number1 =int(line.split(' ')[2])\n",
    "        number2= int(line.split(' ')[3].replace(\"\\n\",''))\n",
    "        region_ranges.append([number1,number2])\n",
    "\n",
    "for region in IndexesOfOffgene:\n",
    "    rangeofregion = region_ranges[IndexesOfOffgene.index(region)]\n",
    "    Allnumbers = list(range(rangeofregion[0],rangeofregion[1]))\n",
    "    offgeneregion = list(set(Allnumbers) - set(region))\n",
    "    offgene.append(offgeneregion)\n",
    "\n",
    "ALLCDS = seq_regions\n",
    "# konec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(promotors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import fastaparser\n",
    "\n",
    "\n",
    "def Change_base(i):\n",
    "    switcher = {\n",
    "        \"A\": \"T\",\n",
    "        \"T\": \"A\",\n",
    "        \"C\": \"G\",\n",
    "        \"G\": \"C\",\n",
    "        \"a\": \"t\",\n",
    "        \"t\": \"a\",\n",
    "        \"c\": \"g\",\n",
    "        \"g\": \"c\",\n",
    "        }   \n",
    "    return switcher.get(i,i)\n",
    "\n",
    "\n",
    "\n",
    "def rotate_seq(dataToRotate: str):\n",
    "    reversedstring = dataToRotate[::-1]\n",
    "    new_string = []\n",
    "    for base in reversedstring :\n",
    "        new_base = Change_base(base)\n",
    "        new_string.append(new_base)\n",
    "    new_data = \"\".join(new_string)\n",
    "    return new_data\n",
    "\n",
    "FnaCDS = []\n",
    "FnaOff =[]\n",
    "Fnaprom = []\n",
    "with open(path_to_fna, \"r\") as rf:\n",
    "    parser = fastaparser.Reader(rf, parse_method=\"quick\")\n",
    "    region = 0 \n",
    "    for data in parser:\n",
    "        header = data.header\n",
    "        sequence = data.sequence\n",
    "        for CDS in AllCDS[region] :\n",
    "            string = sequence[int(CDS[0])-1:int(CDS[1])]\n",
    "            if CDS[3] == \"-\":\n",
    "                string = rotate_seq(string)\n",
    "            FnaCDS.append([string,CDS[5],CDS[3]])\n",
    "        for promotor in promotors[region]:\n",
    "            string = sequence[int(promotor[0])-1:int(promotor[1])]\n",
    "            if promotor[2] == \"-\":\n",
    "                string = rotate_seq(string)\n",
    "            Fnaprom.append([string,promotor[4],promotor[2],promotor[3]])\n",
    "        string = []\n",
    "        for off in offgene[region] :\n",
    "            try:\n",
    "                string.append(sequence[off-1])\n",
    "            except:\n",
    "                continue\n",
    "        string = ''.join(string)\n",
    "        FnaOff.append([string,header.split(' ')[0]])\n",
    "        region = region + 1\n",
    "\n",
    "pathtodata = \"E:/VYSOKA/BAKA/data/\"\n",
    "with open(pathtodata + \"_Candida_auris/\" + \"_Candida_auris_promotor.txt\",\"w\") as f:\n",
    "    for prom in Fnaprom:\n",
    "        f.write(\"\\n\" + prom[1]+ '   ' + prom[2]+ '   ' + str(prom[3]))\n",
    "        line = prom[0]\n",
    "        n = 80\n",
    "        newline = [line[i:i+n] for i in range(0, len(line), n)]\n",
    "        for line in newline:\n",
    "            f.write(\"\\n\" + line)\n",
    "\n",
    "with open(pathtodata + \"_Candida_auris/\" + \"_Candida_auris_CDS.txt\",\"w\") as f:\n",
    "    for CDS in FnaCDS:\n",
    "        f.write(\"\\n\" + CDS[1]+ '   ' + CDS[2])\n",
    "        line = CDS[0]\n",
    "        n = 80\n",
    "        newline = [line[i:i+n] for i in range(0, len(line), n)]\n",
    "        for line in newline:\n",
    "            f.write(\"\\n\" + line)\n",
    "\n",
    "with open(pathtodata + \"_Candida_auris/\" + \"_Candida_auris_Offgene.txt\",\"w\") as f:\n",
    "    for offs in FnaOff:\n",
    "        f.write(\"\\n\" + offs[1])\n",
    "        line = offs[0]\n",
    "        n = 80\n",
    "        newline = [line[i:i+n] for i in range(0, len(line), n)]\n",
    "        for line in newline:\n",
    "            f.write(\"\\n\" + line)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(off)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for off in offgene:\n",
    "    print(len(off))\n",
    "\n",
    "4049*80"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "dir_list = os.listdir(\"E:/VYSOKA/BAKA/data\")\n",
    "dir_list2 = os.listdir(\"E:/VYSOKA/BAKA/data/\" + dir_list[0])\n",
    "\n",
    "print(dir_list2)\n",
    "\n",
    "for dir in dir_list:\n",
    "    print(dir)\n",
    "    os.remove(\"E:/VYSOKA/BAKA/data/\" + dir +\"/\" + dir +\"_CDS.txt\")\n",
    "    os.remove(\"E:/VYSOKA/BAKA/data/\" + dir +\"/\" + dir +\"_Offgene.txt\")\n",
    "    os.remove(\"E:/VYSOKA/BAKA/data/\" + dir +\"/\" + dir +\"_promotor.txt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "dir_list = os.listdir(\"E:/VYSOKA/BAKA/data\")\n",
    "dir_list2 = os.listdir(\"E:/VYSOKA/BAKA/data/\" + dir_list[102])\n",
    "print(dir_list[102])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "promotor_plus = []\n",
    "promotor_minus = []\n",
    "\n",
    "for region in CDS_plus:\n",
    "    promotors_in_region = []\n",
    "    for cds in region:\n",
    "        promotors_in_region.append([int(cds[0])-1000,int(cds[0])-1])\n",
    "    promotor_plus.append(promotors_in_region)\n",
    "\n",
    "for region in CDS_minus:\n",
    "    promotors_in_region = []\n",
    "    for cds in region:\n",
    "        promotors_in_region.append([int(cds[1])+1,int(cds[1])+1000])\n",
    "    promotor_minus.append(promotors_in_region)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_promotor_plus = promotor_plus\n",
    "old_promotor_minus = promotor_minus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(CDS_plus[0][0:10])\n",
    "print(promotor_plus[0][0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for region in promotor_plus :\n",
    "    for promotor in region:\n",
    "        if region.index(promotor)  < 1 : continue\n",
    "        if promotor[0] <= int(CDS_plus[promotor_plus.index(region)][region.index(promotor)-1][1]) :\n",
    "            print(promotor)\n",
    "            promotor[0] = int(CDS_plus[promotor_plus.index(region)][region.index(promotor)-1][1]) + 1\n",
    "            print(promotor)\n",
    "            print('-----------------')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for region in promotor_minus :\n",
    "    for promotor in region:\n",
    "        if region.index(promotor)+1  == len(CDS_minus[promotor_minus.index(region)]): \n",
    "            continue\n",
    "        if promotor[1] >= int(CDS_minus[promotor_minus.index(region)][region.index(promotor)+1][0]) :\n",
    "            print(promotor)\n",
    "            promotor[1] = int(CDS_minus[promotor_minus.index(region)][region.index(promotor)+1][0]) - 1\n",
    "            print(promotor)\n",
    "            print('-----------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(CDS_plus[0][0:10])\n",
    "print(promotor_plus[0][0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Numbers = []\n",
    "\n",
    "for region in CDS_plus:\n",
    "    nbr_in_region = [0]\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        nbr_in_region =nbr_in_region + list(range(number1,number2))\n",
    "    Numbers.append(nbr_in_region)\n",
    "\n",
    "for region in CDS_minus :\n",
    "    nbr_in_region = [0]\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        nbr_in_region = nbr_in_region + list(range(number1,number2))\n",
    "    Numbers[CDS_minus.index(region)] = Numbers[CDS_minus.index(region)] + nbr_in_region\n",
    "\n",
    "for region in promotor_minus :\n",
    "    nbr_in_region = [0]\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        nbr_in_region = nbr_in_region + list(range(number1,number2))\n",
    "    Numbers[promotor_minus.index(region)] = Numbers[promotor_minus.index(region)] + nbr_in_region\n",
    "\n",
    "for region in promotor_plus :\n",
    "    nbr_in_region = [0]\n",
    "    for numbers in region :\n",
    "        number1 = int(numbers[0])\n",
    "        number2 = int(numbers[1]) \n",
    "        nbr_in_region =nbr_in_region + list(range(number1,number2))\n",
    "    Numbers[promotor_plus.index(region)] = Numbers[promotor_plus.index(region)] + nbr_in_region\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_ranges = []\n",
    "offgene = []\n",
    "\n",
    "for line in lines_gff:\n",
    "    if \"sequence-region\" in line:\n",
    "        number1 =int(line.split(' ')[2])\n",
    "        number2= int(line.split(' ')[3].replace(\"\\n\",''))\n",
    "        region_ranges.append([number1,number2])\n",
    "\n",
    "for region in Numbers:\n",
    "    rangeofregion = region_ranges[Numbers.index(region)]\n",
    "    Allnumbers = list(range(rangeofregion[0],rangeofregion[1]))\n",
    "    offgeneregion = list(set(Allnumbers) - set(region))\n",
    "    offgene.append(offgeneregion)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stri = \"GACTGAGAAGTAACGAAGC\"\n",
    "\n",
    "def Change_base(i):\n",
    "    switcher = {\n",
    "        \"A\": \"T\",\n",
    "        \"T\": \"A\",\n",
    "        \"C\": \"G\",\n",
    "        \"G\": \"C\",\n",
    "        \"a\": \"t\",\n",
    "        \"t\": \"a\",\n",
    "        \"c\": \"g\",\n",
    "        \"g\": \"c\",\n",
    "        }   \n",
    "    return switcher.get(i,i)\n",
    "\n",
    "\n",
    "\n",
    "def rotate_seq(dataToRotate: str):\n",
    "    print(dataToRotate)\n",
    "    reversedstring = dataToRotate[::-1]\n",
    "    print(reversedstring)\n",
    "    new_string = []\n",
    "    for base in reversedstring :\n",
    "        new_base = Change_base(base)\n",
    "        new_string.append(new_base)\n",
    "    new_data = \"\".join(new_string)\n",
    "    return new_data\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "newstri = rotate_seq(stri)\n",
    "print(newstri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['E:/VYSOKA/BAKA/data/Candida_albicans/Calb1.ctss', 'E:/VYSOKA/BAKA/data/Candida_albicans/GCF_000182965.3_ASM18296v3_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Kluyveromyces_lactis/Klac1.ctss', 'E:/VYSOKA/BAKA/data/Kluyveromyces_lactis/GCF_000002515.2_ASM251v1_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Naumovozyma_castellii/Ncas1.ctss', 'E:/VYSOKA/BAKA/data/Naumovozyma_castellii/GCF_000237345.1_ASM23734v1_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Saccharomyces_paradoxus/Spar1.ctss', 'E:/VYSOKA/BAKA/data/Saccharomyces_paradoxus/GCF_002079055.1_ASM207905v1_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Schizosaccharomyces_japonicus/Sjap1.ctss', 'E:/VYSOKA/BAKA/data/Schizosaccharomyces_japonicus/GCF_000149845.2_SJ5_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Schizosaccharomyces_pombe/Spom1.ctss', 'E:/VYSOKA/BAKA/data/Schizosaccharomyces_pombe/GCF_000002945.1_ASM294v2_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Yarrowia_lipolytica/Ylip1.ctss', 'E:/VYSOKA/BAKA/data/Yarrowia_lipolytica/GCF_000002525.2_ASM252v1_genomic.fna']\n"
     ]
    }
   ],
   "source": [
    "import fastaparser\n",
    "\n",
    "\n",
    "def Change_base(i):\n",
    "    switcher = {\n",
    "        \"A\": \"T\",\n",
    "        \"T\": \"A\",\n",
    "        \"C\": \"G\",\n",
    "        \"G\": \"C\",\n",
    "        \"a\": \"t\",\n",
    "        \"t\": \"a\",\n",
    "        \"c\": \"g\",\n",
    "        \"g\": \"c\",\n",
    "        }   \n",
    "    return switcher.get(i,i)\n",
    "\n",
    "\n",
    "\n",
    "def rotate_seq(dataToRotate: str):\n",
    "    reversedstring = dataToRotate[::-1]\n",
    "    new_string = []\n",
    "    for base in reversedstring :\n",
    "        new_base = Change_base(base)\n",
    "        new_string.append(new_base)\n",
    "    new_data = \"\".join(new_string)\n",
    "    return new_data\n",
    "\n",
    "\n",
    "import os\n",
    "dir_list = os.listdir(\"E:/VYSOKA/BAKA/data\")\n",
    "dir_list2 = os.listdir(\"E:/VYSOKA/BAKA/data/Candida_albicans/\")\n",
    "\n",
    "def find_dirs():\n",
    "    dir_list = os.listdir(\"E:/VYSOKA/BAKA/data\")\n",
    "    for dir in dir_list:\n",
    "        dir_list2 = os.listdir(\"E:/VYSOKA/BAKA/data/\" +dir)\n",
    "        for file in dir_list2:\n",
    "            if \"ctss\" not in file.split('.')[1] : continue\n",
    "            if \"2\" in file : continue\n",
    "            for file2 in dir_list2:\n",
    "                if \"fna\" != file2.split(\".\")[-1] : continue\n",
    "                fnafile = file2\n",
    "            yield([\"E:/VYSOKA/BAKA/data/\" +dir +\"/\" + file,\"E:/VYSOKA/BAKA/data/\" +dir +\"/\" + fnafile])\n",
    "    \n",
    "\n",
    "def tss_data(d):\n",
    "    tssdata = {}\n",
    "    seqence = {}\n",
    "    for file in d:\n",
    "        print(file)\n",
    "        path_to_fna = file[1]\n",
    "        path_to_tss = file[0]\n",
    "\n",
    "        with open(path_to_tss,\"r\") as f:\n",
    "            lines_tss = f.readlines()\n",
    "        \n",
    "        for line in lines_tss:\n",
    "            l = line.split(\"\\t\")\n",
    "            if l[0] not in tssdata:\n",
    "                tssdata[l[0]] = [[l[1],l[2]]]\n",
    "                continue\n",
    "        \n",
    "            tssdata[l[0]].append([l[1],l[2]])\n",
    "\n",
    "        with open(path_to_fna, \"r\") as rf:\n",
    "            parser = fastaparser.Reader(rf, parse_method=\"quick\")\n",
    "            for data in parser:\n",
    "                dataheader = data.header.split(\" \")[0].replace('>', '')\n",
    "                if dataheader not in tssdata : continue\n",
    "                if dataheader not in seqence:\n",
    "                    seqence[dataheader] = []\n",
    "                \n",
    "                emptysten = []\n",
    "                for tss in tssdata[dataheader]:\n",
    "                    if tss[1] == \"-\":\n",
    "                        string = data.sequence[int(tss[0])-700:int(tss[0])+300]\n",
    "                        if len(string) == 0 : emptysten.append(tss)\n",
    "                        string = rotate_seq(string)\n",
    "                        string = [string,tss[1]]\n",
    "                        seqence[dataheader].append(string)\n",
    "                        continue\n",
    "                    \n",
    "                    string = [data.sequence[int(tss[0])-299:int(tss[0])+701],tss[1]]    \n",
    "                    seqence[dataheader].append(string)\n",
    "\n",
    "\n",
    "        newFilePath = file[0].split(\"/\")\n",
    "        newFilePath[-1] = newFilePath[-2] +\"_promotor_tss.txt\"\n",
    "        newFilePath = (\"/\").join(newFilePath)\n",
    "        with open(newFilePath, \"w\") as f :\n",
    "            for key in seqence.keys():\n",
    "                for data in seqence[key]:\n",
    "                    f.write(\"\\n\" +key+ '   ' + data[1])\n",
    "                    line = data[0]\n",
    "                    n = 80\n",
    "                    newline = [line[i:i+n] for i in range(0, len(line), n)]\n",
    "                    for line in newline:\n",
    "                        f.write(\"\\n\" + line)\n",
    "                    \n",
    "            \n",
    "            \n",
    "\n",
    "\n",
    "        \n",
    "\n",
    "                \n",
    "\n",
    "        \n",
    "        \n",
    "        \n",
    "d=find_dirs()          \n",
    "tss_data(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['E:/VYSOKA/BAKA/data/Candida_albicans/Calb1.ctss', 'E:/VYSOKA/BAKA/data/Candida_albicans/GCF_000182965.3_ASM18296v3_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Kluyveromyces_lactis/Klac1.ctss', 'E:/VYSOKA/BAKA/data/Kluyveromyces_lactis/GCF_000002515.2_ASM251v1_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Naumovozyma_castellii/Ncas1.ctss', 'E:/VYSOKA/BAKA/data/Naumovozyma_castellii/GCF_000237345.1_ASM23734v1_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Saccharomyces_paradoxus/Spar1.ctss', 'E:/VYSOKA/BAKA/data/Saccharomyces_paradoxus/GCF_002079055.1_ASM207905v1_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Schizosaccharomyces_japonicus/Sjap1.ctss', 'E:/VYSOKA/BAKA/data/Schizosaccharomyces_japonicus/GCF_000149845.2_SJ5_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Schizosaccharomyces_pombe/Spom1.ctss', 'E:/VYSOKA/BAKA/data/Schizosaccharomyces_pombe/GCF_000002945.1_ASM294v2_genomic.fna']\n",
      "['E:/VYSOKA/BAKA/data/Yarrowia_lipolytica/Ylip1.ctss', 'E:/VYSOKA/BAKA/data/Yarrowia_lipolytica/GCF_000002525.2_ASM252v1_genomic.fna']\n"
     ]
    }
   ],
   "source": [
    "d=find_dirs()  \n",
    "for i in d:\n",
    "    print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "19d1d53a962d236aa061289c2ac16dc8e6d9648c89fe79f459ae9a3493bc67b4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
